scripts/can_errgo/error_tui.go

```diff
@@ -0,0 +1,626 @@
+			if currentError != nil{
+				newIndex := findError(currentError[0], newRowValues)
+				m.table.SetCursorAndViewport(newIndex)
+			}
+		}
+		return m, nil
+	// If the message is a tick, check if the last CAN message was received more than the timeout duration ago
+	case TickMsg:
+		if time.Since(m.lastCANTime) > m.t {
+			m.isTimeout = true
+		}
+		return m, tickEvery(m.t)
+	}
+
+	// Update the table with the new message
+	if _, ok := msg.(CANMsg); ok {
+        // Update the mainTable regardless of submenu state
+        m.table, cmd = m.table.Update(msg)
+    }else if m.submenuActive {
+        m.submenuTable, cmd = m.submenuTable.Update(msg)
+    } else {
+        m.table, cmd = m.table.Update(msg)
+    }
+
+	return m, cmd
+}
+
+// KeyMap is a set of keybindings for the application.
+type KeyMap struct {
+	a       key.Binding
+	i       key.Binding
+	s	    key.Binding
+	q       key.Binding
+
+}
+
+// ShortHelp implements the KeyMap interface.
+func (km KeyMap) ShortHelp() []key.Binding {
+	return []key.Binding{km.a, km.i, km.s, km.q}
+}
+
+
+// key bindings for the main table
+func additionalKeyMap() KeyMap {
+	return KeyMap{
+		a: key.NewBinding(
+			key.WithKeys("a"),
+			key.WithHelp("a", "Acknowledge"),
+		),
+		i: key.NewBinding(
+			key.WithKeys("i"),
+			key.WithHelp("i", "Ignore"),
+		),	
+		s: key.NewBinding(
+			key.WithKeys("s"),
+			key.WithHelp("s", "Ignored Menu"),
+		),
+		q: key.NewBinding(
+			key.WithKeys("q", "ctrl+c"),
+			key.WithHelp("q/ctrl+c", "quit"),
+		),
+	}
+}
+
+// key bindings for the submenu
+func subMenuKeyMap() KeyMap {
+	return KeyMap{
+		i: key.NewBinding(
+			key.WithKeys("i"),
+			key.WithHelp("i", "Unignore"),
+		),
+		s: key.NewBinding(
+			key.WithKeys("s"),
+			key.WithHelp("s", "Close Ignored Menu"),
+		),
+		q: key.NewBinding(
+			key.WithKeys("q"),
+			key.WithHelp("q/ctrl+c", "quit"),
+		),
+	}
+}
+
+// View returns the string representation of the model. We could combine and condense, but its clearer to keep it separate.
+func (m model) View() string {
+
+	out := "\n \n \n"
+	// Help menu changes based on the state of the submenu
+	if m.submenuActive{
+		out = 	m.table.HelpView() + "\n" + 
+				m.submenuTable.Help.ShortHelpView(m.submenuKeys.ShortHelp()) + "\n"
+	}else{
+		out = 	m.table.HelpView() + "\n" +
+				m.table.Help.ShortHelpView(m.tableKeys.ShortHelp()) + "\n"	
+	}
+
+	// Timeout warning
+	if m.isTimeout{
+		out = fmt.Sprint("\n\x1b[41m\x1b[37mWarning! Last recorded message was over ",math.Round(time.Since(m.lastCANTime).Seconds()), " seconds ago!\x1b[0m") + "\n" +  out
+	}
+
+	// Render the table 
+	out = out + baseStyle.Render(m.table.View()) + "\n"
+
+	// Render the submenu if it is active
+	if m.submenuActive{
+		out = out + baseStyle.Render(m.submenuTable.View()) + "\n"
+	}
+
+	// Render the box 
+	if m.submenuActive && m.submenuTable.SelectedRow() != nil{
+		m.box.SetText(m.submenuTable.SelectedRow()[0], m.getDescription(m.submenuTable.SelectedRow()[0]))
```

Comment:
This section is reading data from the table. You should have a separate variable to store the index of the currently selected row and use this to directly get the description for that error struct. You also shouldn't be searching for data based on the string name. This should be an index representing the error number.

I think it would be good to rename the `RowValues` struct to `ErrorData`. You should have an array of 64 `ErrorData` structs each with their own name, description, count, recency, etc. When a new error message arrives, update the counts & recencies in this array, but _don't sort this array_. Keep it in order so you always know that error 7 is at position 7. You have some functions which iterate over the table for information (like `getDescription`, `findError`, `resetRowValue`, `getIgnoredRows`) which should all be direct lookups in the ErrorData array.

For example, you should have something in your program which is structured like

```
errors = [
    ErrorData(name="Error0", desc="this is desc of 0", count = 7, recency = 1, ignored=false),
    ErrorData(name="Error1", desc="this is desc of 1", count = 5, recency = 2, ignored=true),
    ErrorData(name="Error2", desc="this is desc of 2", count = 1, recency = 0, ignored=false),
    ErrorData(name="Error3", desc="this is desc of 3", count = 10, recency = 3, ignored=true),
    .... for all 64 errors
]
```

of course this array would be created at runtime, not hardcoded. This should happen right when the program boots.

Then `getDescription` would just be `errors[index].description`.

When a new message comes in, to update the count & recency:

```
for i, msg in range errors:
    msg.recency += 1
    if can_error bit i is set
        msg.recency = 0
        msg.count += 1
```

You will need to figure out how to sort the errors for displaying in the table, but this should be considered after getting the backend structure correct. Think about your program as a whole system with two responsibilities

1. Track the state of each message
2. Sort and display some of the messages in a table

Your logic for 1) should not rely on how 2) works. let's do 1) really well before worrying about how exactly 2) will function